ydb/core/tx/schemeshard/schemeshard_utils.h

```diff
@@ -183,27 +186,48 @@ bool CommonCheck(const TTableDesc& tableDesc, const NKikimrSchemeOp::TIndexCreat
         return false;
     }
 
-    if (!indexKeys.DataColumns.empty() && !AppData()->FeatureFlags.GetEnableDataColumnForIndexTable()) {
-        status = NKikimrScheme::EStatus::StatusPreconditionFailed;
-        error = "It is not allowed to create index with data column";
-        return false;
+    if (!indexKeys.DataColumns.empty()) {
+        if (!AppData()->FeatureFlags.GetEnableDataColumnForIndexTable()) {
+            status = NKikimrScheme::EStatus::StatusPreconditionFailed;
+            error = "It is not allowed to create index with data column";
+            return false;
+        }
     }
 
-    if (!IsCompatibleIndex(baseTableColumns, indexKeys, error)) {
+    if (!IsCompatibleIndex(indexDesc.GetType(), baseTableColumns, indexKeys, error)) {
         status = NKikimrScheme::EStatus::StatusInvalidParameter;
         return false;
     }
 
-    TColumnTypes columnsTypes;
-    if (!ExtractTypes(tableDesc, columnsTypes, error)) {
+    TColumnTypes baseColumnTypes;
+    if (!ExtractTypes(tableDesc, baseColumnTypes, error)) {
         status = NKikimrScheme::EStatus::StatusInvalidParameter;
         return false;
     }
 
-    implTableColumns = CalcTableImplDescription(baseTableColumns, indexKeys);
-    if (!IsCompatibleKeyTypes(columnsTypes, implTableColumns, uniformTable, error)) {
-        status = NKikimrScheme::EStatus::StatusInvalidParameter;
-        return false;
+    implTableColumns = CalcTableImplDescription(indexDesc.GetType(), baseTableColumns, indexKeys);
```

Comment:
Move this line to else block below, in such case modifying CalcTableImplDescription will be unnecessary

Reply:
Agreed. 
It has taken a bit more refactoring, but in general it's better.